### PR TITLE
fix: use correct container name for archived logs

### LINKED ISSUE #13785: UI doesn't show archived logs for containerSet steps
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

I am currently looking into ways to optimize the duration of my workflows. And I am looking at containerSet as well. However, I noticed that accessing logs doesn't work well here. Except for a few seconds after the task was finished, the logs are always empty.

image

But the logs are there, sitting in the artifact repository, waiting to be looked at.

```
multi-step-wt7dp/
multi-step-wt7dp/multi-step-wt7dp-multi-step-2322966654/
multi-step-wt7dp/multi-step-wt7dp-multi-step-2322966654/step-1.log
multi-step-wt7dp/multi-step-wt7dp-multi-step-2322966654/step-2.log
multi-step-wt7dp/multi-step-wt7dp-multi-step-2322966654/step-3.log
```

If I click on the link, below the log area (in the UI), I will see them

image

And yes, I know that the UI specifically states "_Your pod GC settings will delete pods and their logs immediately on completion. Logs may not appear for pods that are deleted._".

However, this doesn't seem to be an issue for any other workflow, and using the podGC strategy `OnPodCompletion` is pretty much standard for all my workflows. Simply put: despite what the UI says, it's usually working anyway, except for containerSet tasks.

Also consider this: Other people (like operators) are supposed to work with this system as well. If they have difficulties to read the logs, or if there is some extra step necessary, which also breaks the UI experience (the logs are not opened in a new tab or something, they will be opened in the same tab the Argo server UI is currently running in, basically meaning, they will leave the Argo UI to see the logs), this will decrease their general acceptance when I put them in front of a new system like Argo workflows.

Steps to reproduce:

1. Deploy the workflow template (artifact repository required)
2. Run the workflow
3. Wait a minute
4. Try to access the logs in the UI

TLDR: The logs are there, Argo can usually show them, even if the pods was already removed, so why does that not work with containerSet?

### Version(s)

v.3.5.11

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: multi-step
  namespace: test
spec:
  entrypoint: entrypoint
  serviceAccountName: test
  artifactGC:
    serviceAccountName: test
    strategy: OnWorkflowDeletion
    forceFinalizerRemoval: true
  podGC:
    strategy: OnPodCompletion
  archiveLogs: true
  activeDeadlineSeconds: 30
  ttlStrategy:
    secondsAfterSuccess: 300
    secondsAfterFailure: 300
  templates:
    - name: entrypoint
      dag:
        tasks:
          - name: multi-step
            template: multi-step
    - name: multi-step
      containerSet:
        containers:
          - name: step-1
            image: docker.io/busybox:1.36
            command: [echo,"step 1"]
          - name: step-2
            dependencies: [step-1]
            image: docker.io/busybox:1.36
            command: [echo,"step 2"]
          - name: step-3
            dependencies: [step-2]
            image: docker.io/busybox:1.36
            command: [echo,"step 3"]
```

### Logs from the workflow controller

```text
time="2024-10-18T11:03:08.780Z" level=info msg="Processing workflow" Phase= ResourceVersion=459976086 namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=info msg="Task-result reconciliation" namespace=test numObjs=0 workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=info msg="Updated phase  -> Running" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=info msg="DAG node multi-step-wt7dp initialized Running" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=warning msg="was unable to obtain the node for multi-step-wt7dp-2322966654, taskName multi-step"
time="2024-10-18T11:03:08.793Z" level=warning msg="was unable to obtain the node for multi-step-wt7dp-2322966654, taskName multi-step"
time="2024-10-18T11:03:08.793Z" level=info msg="All of node multi-step-wt7dp.multi-step dependencies [] completed" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.793Z" level=info msg="Pod node multi-step-wt7dp-2322966654 initialized Pending" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg="Created pod: multi-step-wt7dp.multi-step (multi-step-wt7dp-multi-step-2322966654)" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg="Container node multi-step-wt7dp-4212253844 initialized Pending" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg="Container node multi-step-wt7dp-4262586701 initialized Pending" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg="Container node multi-step-wt7dp-4245809082 initialized Pending" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.801Z" level=info msg=reconcileAgentPod namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:08.812Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=459976091 workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.802Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=459976091 namespace=test workflow=multi-step-wt7d
time="2024-10-18T11:03:14.803Z" level=info msg="Task-result reconciliation" namespace=test numObjs=1 workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="task-result changed" namespace=test nodeID=multi-step-wt7dp-2322966654 workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4212253844 phase Pending -> Succeeded" namespace=test workflow=multi-step-wt7d
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4212253844 finished: 2024-10-18 11:03:14.803299142 +0000 UTC" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4262586701 phase Pending -> Succeeded" namespace=test workflow=multi-step-wt7d
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4262586701 finished: 2024-10-18 11:03:14.803319422 +0000 UTC" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4245809082 phase Pending -> Succeeded" namespace=test workflow=multi-step-wt7d
time="2024-10-18T11:03:14.803Z" level=info msg="node multi-step-wt7dp-4245809082 finished: 2024-10-18 11:03:14.803335402 +0000 UTC" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="node changed" namespace=test new.message= new.phase=Running new.progress=0/1 nodeID=multi-step-wt7dp-2322966654 old.message= old.phase=Pending old.progress=0/1 workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.803Z" level=info msg=reconcileAgentPod namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.809Z" level=info msg="cleaning up pod" action=terminateContainers key=test/multi-step-wt7dp-multi-step-2322966654/terminateContainers
time="2024-10-18T11:03:14.809Z" level=info msg="https://[...]:443/api/v1/namespaces/test/pods/multi-step-wt7dp-multi-step-2322966654/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2024-10-18T11:03:14.814Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=459976476 workflow=multi-step-wt7dp
time="2024-10-18T11:03:14.827Z" level=info msg="signaled container" container=wait error="unable to upgrade connection: container not found (\"wait\")" namespace=test pod=multi-step-wt7dp-multi-step-2322966654 stderr="<nil>" stdout="<nil>"
time="2024-10-18T11:03:20.816Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=459976476 namespace=test workflow=multi-step-wt7d
time="2024-10-18T11:03:20.816Z" level=info msg="Task-result reconciliation" namespace=test numObjs=1 workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.816Z" level=info msg="node changed" namespace=test new.message= new.phase=Succeeded new.progress=0/1 nodeID=multi-step-wt7dp-2322966654 old.message= old.phase=Running old.progress=0/1 workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="Outbound nodes of multi-step-wt7dp set to [multi-step-wt7dp-4245809082]" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="node multi-step-wt7dp phase Running -> Succeeded" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="node multi-step-wt7dp finished: 2024-10-18 11:03:20.817160127 +0000 UTC" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg=reconcileAgentPod namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="Updated phase Running -> Succeeded" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.817Z" level=info msg="Marking workflow completed" namespace=test workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.825Z" level=info msg="Workflow update successful" namespace=test phase=Succeeded resourceVersion=459976763 workflow=multi-step-wt7dp
time="2024-10-18T11:03:20.826Z" level=info msg="Queueing Succeeded workflow test/multi-step-wt7dp for delete in 5m0s due to TTL"
time="2024-10-18T11:03:25.833Z" level=info msg="cleaning up pod" action=deletePod key=test/multi-step-wt7dp-multi-step-2322966654/deletePod
time="2024-10-18T11:03:44.827Z" level=info msg="cleaning up pod" action=killContainers key=test/multi-step-wt7dp-multi-step-2322966654/killContainers
```

### Logs from in your workflow's wait container

```text
No resources found in [...] namespace.
```

### PR DESCRIPTION
**Pull request by gaurang9991**
Fixes #13785 
### Motivation

This PR is to fix the issue when fetching the logs for container set steps. The issue exists due to the wrong container name used in api calls from the frontend. To save the logs we already use file name as containerName.log and a artifact name as containerName-logs

### Modifications

Removed the constraints of main container name. The frontend now uses the correct container name when making API calls to fetch logs for container set steps.  

### Verification

Locally tested by fetching logs for container set steps and verifying the correct container name is used.

Test Workflow Example:
```yaml
metadata:
  name: fantastic-bear
  namespace: argo
  labels:
    example: "true"
spec:
  arguments:
    parameters:
      - name: message
        value: hello argo
  entrypoint: argosay
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            value: "{{workflow.parameters.message}}"
      containerSet:
        containers:
          - name: some-message
            image: argoproj/argosay:v2
            command:
              - /argosay
            args:
              - echo
              - "{{inputs.parameters.message}}"
            resources: {}
          - name: some-message-2
            dependencies:
             - some-message
            image: argoproj/argosay:v2
            command:
              - /argosay
            args:
              - echo
              - "{{inputs.parameters.message}} 2"
            resources: {}
  ttlStrategy:
    secondsAfterCompletion: 300
  podGC:
    strategy: OnPodCompletion
```    
### Container Set
image
image
### Standard template
image

In the images highlighted the name used in containers and api calls. 
- For container set it use the name of the container - some-message
- For standard steps or templates continue to use the main container

### Documentation

Not Required

### AI

None

## Summary by CodeRabbit

- **Bug Fixes**
  - Archived logs are now available for eligible containers in completed workflows, including containers defined in container sets.
  - Log retrieval selects archived logs when available and falls back to live logs for pending or running nodes.
  - Workflow and pod log views automatically select the correct main container for multi-container nodes.
  - Logs remain aligned with the latest workflow state as nodes transition between running and completed.

### CODE REVIEW

File: ui/src/shared/services/workflows-service.ts [hunk 1]
```diff
@@ -263,10 +263,10 @@ export const WorkflowsService = {
             return getLogsFromArtifact();
         }
 
-        // return archived log if main container is finished and has artifact
+        // Return archived logs when the container is finished and has an artifact.
         return from(this.isWorkflowNodePendingOrRunning(workflow, nodeId)).pipe(
             switchMap(isPendingOrRunning => {
-                if (!isPendingOrRunning && hasArtifactLogs(workflow, nodeId, container) && container === 'main') {
+                if (!isPendingOrRunning && hasArtifactLogs(workflow, nodeId, container)) {
```
Comment (Joibel) [inline ui/src/shared/services/workflows-service.ts]: `isWorkflowNodePendingOrRunning` fetches a fresh copy of the workflow to check the node phase. But `hasArtifactLogs` here still reads the old `workflow` object the caller passed in.

If the pod was just deleted and the browser has not yet received the update that adds `step-1-logs` to the node outputs, the fresh copy says "finished" but the old copy says "no artifact". The UI then asks the cluster, gets nothing, and shows an empty pane.

This was possible before for `main` too, but the PR now depends on this path for every containerSet member. The fresh workflow is already in hand inside `isWorkflowNodePendingOrRunning`. Returning it (or the node) from there and using it for `hasArtifactLogs` would remove the problem.

Optional: `hasArtifactLogs` is also checked inside `getContainerLogsFromArtifact`, which throws if it is false. So this could be simplified to one check:

```ts
return isPendingOrRunning
    ? getLogsFromCluster().pipe(catchError(getLogsFromArtifact))
    : getLogsFromArtifact().pipe(catchError(getLogsFromCluster));
```

**Review by Joibel — COMMENTED** [review]
Thanks for tracking this down. The change itself is correct: when you click LOGS on a Container node inside a containerSet, the UI now finds the archived log artifact (for example `step-1-logs`) and shows it.

There is one gap that still leaves #13785 partly broken, plus a missing test. One smaller point is inline on the changed line.

### 1. The LOGS button on the Pod node still shows nothing

`ui/src/workflows/components/workflow-node-info/workflow-node-info.tsx` line 176.

The Pod node has its own LOGS button. It always opens the log viewer with container `main`:

```ts
const showLogs = (x = 'main') => props.onShowContainerLogs(props.node.id, x);
...
<Button onClick={() => showLogs()} icon='bars'>LOGS</Button>
```

A containerSet pod usually has no container called `main`. So this happens:

1. The viewer asks for logs for container `main`.
2. There is no `main-logs` artifact, so the UI asks the cluster instead.
3. The cluster has no container called `main` (or the pod is already deleted). The server drops that error and the stream ends with no lines.
4. The UI only falls back to artifacts when the stream errors, and it did not error. So the pane stays empty and no error is shown.

The user can recover by choosing `step-1` in the container dropdown, but they have to know to do that. This was already broken before your PR, but it is the first thing someone hitting #13785 will click.

Suggested fix: when opening the viewer, pick the first containerSet container if the template has one, otherwise `main`. The same `main` default is also at `workflow-node-info.tsx:339` and `workflow-details.tsx:50`, so a small shared helper would keep the three places in step. The Go side already does this in `Template.GetMainContainerNames`.

### 2. Any container named `X` now matches an artifact named `X-logs`

`ui/src/shared/services/workflows-service.ts` line 25.

`hasArtifactLogs` only checks that an artifact called `<container>-logs` exists. Before, this check only ran for `main`. Now it runs for every container in the dropdown, including sidecars, init and wait.

Example: a template has a sidecar called `nginx` and a normal output artifact called `nginx-logs` (a directory, saved as a tgz). After this PR, choosing `nginx` in the viewer downloads the tgz and prints its bytes as log lines.

A simple guard would be to only prefer artifacts for the template's main containers (the containerSet names, or `main`).

### 3. Please add a test

`ui/src/shared/services/workflow-service.test.ts` already exists. Following the pattern in `cron-workflow-service.test.ts` (mock `./requests`, spy on the functions), a short test could check that a finished node with a `step-1-logs` artifact takes the artifact path. That would stop someone adding the `main` check back by accident.

### Smaller, pre-existing, no action needed

- `workflow-logs-viewer.tsx:171`: the node id is fixed to the node the panel was opened from. If the user switches the pod dropdown to "All" or another pod, the archived log shown is still from the original node. This now affects containerSet members too, not just `main`.
- `workflows-service.ts:253`: the filter text is used directly as a regex. A bad pattern like `error[` throws, the error is swallowed, and the pane is empty. Same result as before the PR.

File: ui/src/shared/services/workflows-service.ts (hunk 1 above)
Comment (gaurang9991) [reply ui/src/shared/services/workflows-service.ts]: following your description and suggestions, i have made some changes 

- If pod logs are viewed then it picks the first container in container set otherwise main
- hasArtifactLogs now takes the updated workflow fetched from server
- logs are fetch for only container set and main container not for sidecar
- added test for to check the getContainerLogs

Now the only edge case i see is the part where the container name and output artifact name is same. even though they are stored in different paths but the based on how the is api it seems we have no way to distinguish it. for now i expect that users to be mindful of this.

you can verify the changes and let me know if they are good.

File: ui/src/shared/services/workflows-service.ts [hunk 2]
```diff
@@ -22,7 +23,7 @@ function isNodePendingOrRunning(node: NodeStatus) {
 
 function hasArtifactLogs(workflow: Workflow, nodeId: string, container: string) {
     const node = workflow.status.nodes[nodeId];
-    return node?.outputs?.artifacts?.some(a => a.name === `${container}-logs`);
+    return !!node && getMainContainerNames(workflow, node).includes(container) && node.outputs?.artifacts?.some(a => a.name === `${container}-logs`);
```
Comment (Joibel) [inline ui/src/shared/services/workflows-service.ts:26]: The new `getMainContainerNames(...).includes(container)` check makes this depend on template resolution, and it fails closed when that doesn't work.

Inline templates are the case I hit on paper. For a DAG task or step with `inline:`, the controller stores `templateName: ""` (`operator.go` sets `TemplateName: orgTmpl.GetTemplateName()`, and that's empty for inline). `getResolvedTemplates` then logs "StoredTemplate  not found" and returns `undefined`, `getMainContainerNames` falls back to `['main']`, and `step-1` gets rejected. So #13785 is still broken for inline containerSets.

It also changes the deleted/archived path. `getContainerLogsFromArtifact` goes through `hasArtifactLogs` too, and on main that was only a name check, so containerSet logs for workflows no longer in the cluster already worked there. With this change they need template resolution to succeed first.

The executor only writes `<name>-logs` for `Template.GetMainContainerNames()` (`executor.go`), so if the artifact is there the container is already a main one. I'd either keep this as a name-only check, or skip the membership check when the template can't be resolved rather than defaulting to `['main']`.

File: ui/src/shared/services/workflow-service.test.ts [hunk 3]
```diff
@@ -40,4 +47,96 @@ describe('workflow service', () => {
...
+                        name: 'container-set',
+                        containerSet: {containers: [{name: 'step-1'}]}
+                    }
+                ]
+            },
+            status: {
+                nodes: {
+                    'test-node': {
+                        id: 'test-node',
+                        phase: 'Running',
```
Comment (Joibel) [inline ui/src/shared/services/workflow-service.test.ts:125]: This one doesn't reach the new check. The node is `phase: 'Running'`, so `isPendingOrRunning` is true and we go straight to the cluster before `hasArtifactLogs` runs. It passes on main, and it still passes with the `.includes(container)` check removed, so nothing currently pins that rule.

Could you make it a `Succeeded` node with a `<name>-logs` artifact for a container that isn't in the containerSet, and assert we go to the cluster? An inline-template case would be worth having too (see my comment on `hasArtifactLogs`).

A couple of smaller things in this file:
- `getMainContainerNames` has no tests of its own. The no-node case, an unresolvable template, and a containerSet with several containers (callers take `[0]`) would all be cheap to cover.
- The first test passes `workflow('hello-world', ...)` but the `get` mock returns `test-workflow-1`, and nothing checks what `get` was called with. Matching names plus `expect(get).toHaveBeenCalledWith('argo', 'hello-world')` would pin the refresh.

File: ui/src/shared/services/workflows-service.ts [hunk 4]
```diff
@@ -255,21 +256,21 @@ export const WorkflowsService = {
...
         if (!isWorkflowInCluster(workflow)) {
             return getLogsFromArtifact();
         }
 
-        // return archived log if main container is finished and has artifact
-        return from(this.isWorkflowNodePendingOrRunning(workflow, nodeId)).pipe(
-            switchMap(isPendingOrRunning => {
-                if (!isPendingOrRunning && hasArtifactLogs(workflow, nodeId, container) && container === 'main') {
-                    return getLogsFromArtifact().pipe(catchError(getLogsFromCluster));
+        // Return archived logs when the container is finished and has an artifact.
```
Comment (Joibel) [inline ui/src/shared/services/workflows-service.ts:267]: Small one: `isPendingOrRunning` is the node's phase, not the container's. In a containerSet one container can finish while the node is still Running, so "when the container is finished" isn't quite what this checks. Maybe "Return archived logs when the node has finished and has a `<container>-logs` artifact for one of its main containers."

File: ui/src/shared/services/workflows-service.ts (hunk 2 above)
Comment (gaurang9991) [reply ui/src/shared/services/workflows-service.ts]: i get the incline container part and reverted this to name check only. But the getMainContainerNames still fallback to ```main``` in case template is not resolved. so this mean when check the pods logs for inline container it pick main as default. this will still if we try to get logs for individual container step. 

I feel we also look into resolving inline container properly in the UI but that is out of scope for this.

File: ui/src/shared/services/workflow-service.test.ts (hunk 3 above)
Comment (gaurang9991) [reply ui/src/shared/services/workflow-service.test.ts]: fixed the tests. 
i did not get test for getMainContainerNames as it is outside the workflowservice. and the use case you asked is with shows logs. do you want me to add it there ?

also i added case inline workflow but in essence it works as the same as others
